### PR TITLE
Updated the markdown for Task1 in Exercise1 penguin

### PR DESCRIPTION
**Pull request by surbhigoel77**

### CODE REVIEW

File: exercises/01_penguin_classification.ipynb [hunk 1]
```diff
@@ -23,20 +23,23 @@
...
-    "- Call this function, which returns a single object, and assign it to the variable ``data``.\n",
-    "  - Print ``data`` and recognise that ``load_penguins`` has returned a ``pandas.DataFrame``.\n",
-    "- Consider which features it might make sense to use in order to classify the species of the penguins.\n",
-    "  - You can print the column titles using ``pd.DataFrame.keys()``\n",
-    "  - You can also obtain useful information using ``pd.DataFrame.Series.describe()``"
+    "- Call this function, which returns a single object in the form of a ``pandas.DataFrame``, and assign it to the variable ``data``.\n",
+    "  - Print ``data`` and recognise that ``load_penguins`` has returned the dataframe.\n",
+    "- Analyse which features it might make sense to use in order to classify the species of the penguins.\n",
+    "  - You can print the column names using ``pd.DataFrame.keys()``\n",
+    "  - You can also obtain useful statical information on the dataset using ``pd.DataFrame.Series.describe()``"
```
Comment (ma595) [inline exercises/01_penguin_classification.ipynb:30]: statistical

File: worked-solutions/01_penguin_classification_solutions.ipynb [hunk 2]
```diff
@@ -108,23 +108,25 @@
    "source": [
     "### Task 2: creating a ``torch.utils.data.Dataset``\n",
     "\n",
-    "All PyTorch dataset objects are subclasses of the ``torch.utils.data.Dataset`` class. To make a custom dataset, create a class which inherits from the ``Dataset`` class, implement some methods (the Python magic (or dunder) methods ``__len__`` and ``__getitem__``) and supply some data.\n",
+    "To be able to use Pytorch functionalities, we need to make the dataset compatible with Pytorch. We do it using PyTorch's Dataset class called ``torch.utils.data.Dataset``. \n",
```
Comment (ma595) [inline worked-solutions/01_penguin_classification_solutions.ipynb:111]: *P*y*T*orch

Comment (ma595) [reply worked-solutions/01_penguin_classification_solutions.ipynb:111]: Why change this comment? We need to talk through dunder methods.

File: exercises/01_penguin_classification.ipynb [hunk 3]
```diff
@@ -23,20 +23,23 @@
     "### Task 1: look at the data\n",
     "In the following code block, we import the ``load_penguins`` function from the ``palmerpenguins`` package.\n",
     "\n",
-    "- Call this function, which returns a single object, and assign it to the variable ``data``.\n",
-    "  - Print ``data`` and recognise that ``load_penguins`` has returned a ``pandas.DataFrame``.\n",
-    "- Consider which features it might make sense to use in order to classify the species of the penguins.\n",
-    "  - You can print the column titles using ``pd.DataFrame.keys()``\n",
-    "  - You can also obtain useful information using ``pd.DataFrame.Series.describe()``"
+    "- Call this function, which returns a single object in the form of a ``pandas.DataFrame``, and assign it to the variable ``data``.\n",
```
Comment (ma595) [inline exercises/01_penguin_classification.ipynb:26]: I would remove 'in the form of' and just say ',a ``pandas.DataFrame``'

**Review by ma595 — CHANGES_REQUESTED** [review]
Some minor comments.

File: worked-solutions/01_penguin_classification_solutions.ipynb [hunk 4]
```diff
@@ -520,17 +563,20 @@
...
-    "  - For those of you who've studied physics, you may be remininded of the partition function in thermodynamics.\n",
     "  - This activation function is good for classifcation when the result is one of ``A or B or C``.\n",
     "    - It's bad if you even want to assign two classification to one images—say a photo of a dog _and_ a cat.\n",
     "  - It turns the raw outputs, or logits, into \"psuedo probabilities\", and we take our prediction to be the most probable class.\n",
     "\n",
+    "\n",
+    "- The purpose of training loop is that for each epoch, all the mini-batches are sequentially passed through the model to make predictions, compute the loss, and update the model parameters and repeat it until we receive a satisfactory performance.\n",
+    "- The validation loop works similarly, but the purpose is to validate the model training's performance without any weight updates. Hence, we are not calculating any gradient during validation. This is made sure using the ``no_grad`` decorator with validation loop.\n",
```
Comment (ma595) [inline worked-solutions/01_penguin_classification_solutions.ipynb:572]: The `@no_grad` decorator disables gradient tracking in the function. We could point out that it is optional. 

We also don't need gradients in the batch_level_accuracy() function.

File: exercises/01_penguin_classification.ipynb (hunk 1 above)
Comment (ma595) [reply exercises/01_penguin_classification.ipynb:30]: I think 'Consider' is probably fine here, but either is fine.

**Review by ma595 — CHANGES_REQUESTED** [review]
These changes are helpful to understand the material. I've made a few comments here and there. 

- Your github profile configuration is a bit messed up. Can you please fix this.
- You've probably already been done this, but when commiting the solutions, restart the kernel, clear the outputs and run everything through once more. 
- I think adding some steps (as python comments) in task 9, task 10 and task 11 would make it easier to complete. See how I've done this in PR #71 for the first set of tasks. Maybe we should open another PR for these changes? 

Some other things worth mentioning either in the session itself or perhaps in a later PR:

- What are the implications of the batch size (i.e. too large and model is prone to overfitting / too small and we get inefficient use of hardware)
- Task 6: We could mention feature scaling / normalisation using batch norm. 
- Optimiser (could include the lr here).
- What `model.train()` and `model.eval()` are doing. Just noticed there's a mention of model.train() and model.eval() already. 
- Whether the train & valid routines lead to good code or not. (repetition - often this is packaged in a training class / alternatively use pytorch lightning.

**Comment by ma595** [comment]
This will also need to be rebased on #71

**Comment by surbhigoel77** [comment]
Thanks for the comments @ma595. I have  made the changes as highlighted by you.

File: worked-solutions/01_penguin_classification_solutions.ipynb [hunk 5]
```diff
@@ -212,21 +246,23 @@
...
     "\n",
-    "These transforms can be very useful for mapping between file paths and tensors of images, etc.\n",
+    "These transforms are very useful for converting file paths to tensors of images and for performing other necessary preprocessing tasks.\n",
     "\n",
-    "- Note: here we create a training and validation set.\n",
-    "    - We allow the model to learn directly from the training set—i.e. we fit the function to these data.\n",
-    "    - During training, we monitor the model's performance on the validation set in order to check how it's doing on unseen data. Normally, people use the validation performance to determine when to stop the training process.\n",
+    "- Note: Here we create a training and validation set.\n",
+    "    - We allow the model to learn from the training setn i.e. we fit the function to these data.\n",
```
[the remark this answers is not in the document]
Comment (ma595) [reply worked-solutions/01_penguin_classification_solutions.ipynb:255]: has this been fixed?

File: worked-solutions/01_penguin_classification_solutions.ipynb [hunk 6]
```diff
@@ -323,19 +356,19 @@
...
     "\n",
     "\n",
-    "Note: we are going to use batch normalisation layers in our network, which don't work if the batch size is one. This can happen on the last batch, if we don't choose a batch size that evenly divides the number of items in the data set. To avoid this, we can set the ``drop_last`` argument to ``True``. The last batch, which will be of size ``len(data_set) % batch_size`` gets dropped, and the data are reshuffled. This is only relevant during the training process - validation will use population statistics."
+    "Note: we are going to use batch normalisation in our network, which doesn't work if the batch size is one. This can happen on the last batch, if we don't choose a batch size that evenly divides the number of items in the data set. To avoid this, we can set the ``drop_last`` argument to ``True``. The last batch, which will be of size ``len(data_set) % batch_size`` gets dropped, and the data are reshuffled. This is only relevant during the training process - validation will use population statistics."
```
Comment (ma595) [inline worked-solutions/01_penguin_classification_solutions.ipynb:366]: What is the reason for changing from`batch normalisation layers` to `batch normalisation`?

File: worked-solutions/01_penguin_classification_solutions.ipynb [hunk 7]
```diff
@@ -520,17 +558,20 @@
     "\n",
     "- Before we jump in and write these loops, we must first choose an activation function to apply to the model's outputs.\n",
     "  - Here we are going to use the softmax activation function: see [the PyTorch docs](https://pytorch.org/docs/stable/generated/torch.nn.Softmax.html).\n",
-    "  - For those of you who've studied physics, you may be remininded of the partition function in thermodynamics.\n",
```
Comment (ma595) [inline worked-solutions/01_penguin_classification_solutions.ipynb:523]: I quite like this comment and would like it to be kept.

File: worked-solutions/01_penguin_classification_solutions.ipynb [hunk 8]
```diff
@@ -473,14 +510,16 @@
    "source": [
     "### Task 7: Selecting a loss function\n",
     "\n",
-    "- Binary cross-entropy is about the most common loss function for classification.\n",
```
[the remark this answers is not in the document]
Comment (ma595) [reply worked-solutions/01_penguin_classification_solutions.ipynb:476]: Actually, I agree, your change makes more sense.

**Review by ma595 — COMMENTED** [review]
Missed a few other changes during the first review, and resolved some other requests as they were really good additions.

File: worked-solutions/01_penguin_classification_solutions.ipynb (hunk 7 above)
Comment (surbhigoel77) [reply worked-solutions/01_penguin_classification_solutions.ipynb:523]: Corrected

**Comment by jatkinson1000** [comment]
@surbhigoel77 and @ma595 checking on this again.
Can you provide an update?

Not sure if this will appear at the summer school, but would be good to do some cleaning/prep with it on the horizon.
